compiler/noirc_frontend/src/node_interner.rs

```diff
@@ -868,11 +872,44 @@ impl NodeInterner {
         self.trait_implementations.get(key).cloned()
     }
 
-    pub fn add_trait_implementation(&mut self, key: &TraitImplKey, trait_impl: Shared<TraitImpl>) {
+    pub fn add_trait_implementation(
+        &mut self,
+        key: &TraitImplKey,
+        trait_impl: Shared<TraitImpl>,
+    ) -> bool {
         self.trait_implementations.insert(key.clone(), trait_impl.clone());
-
-        for func_id in &trait_impl.borrow().methods {
-            self.add_method(&key.typ, self.function_name(func_id).to_owned(), *func_id);
+        match &key.typ {
+            Type::Struct(struct_type, _generics) => {
+                for func_id in &trait_impl.borrow().methods {
+                    let method_name = self.function_name(func_id).to_owned();
+                    let key = (struct_type.borrow().id, method_name);
+                    self.struct_methods.insert(key, *func_id);
+                }
+                true
+            }
+            Type::FieldElement
+            | Type::Unit
+            | Type::Array(..)
+            | Type::Integer(..)
+            | Type::Bool
+            | Type::Tuple(..)
+            | Type::String(..)
+            | Type::FmtString(..) => {
+                for func_id in &trait_impl.borrow().methods {
+                    let method_name = self.function_name(func_id).to_owned();
+                    let key = (key.typ.clone(), method_name);
+                    self.primitive_trait_impls.insert(key, *func_id);
+                }
+                true
+            }
+            Type::TypeVariable(..)
+            | Type::NamedGeneric(..)
+            | Type::Function(..)
```

Comment:
We should allow implementing traits on function types. NamedGenerics will also eventually be possible once we support generics `impl<T> Foo for T` but it's fine not to include these until we do.

Reply:
I added an note in comment, maybe we should create an github issue for this ?

Reply:
Since it looks like you already made #2998, I'll resolve this